src/CoxPH/CenterAndScale.hs

```diff
@@ -1,115 +1,111 @@
-                   -> V.Vector ScaleCovariateIndicator
-                   -- ^ Indicators for whether each of the covariates should be centered and scaled. It is required that the length of this vector is equal to the number of vectors in the design matrix.
-                   -> Either T.Text (V.Vector (VS.Vector Double, Double))
+-- | Conditionally center and scale a design matrix
+centerAndScaleCovs ::
+    -- | The covariate vectors comprising the design matrix. The covariate vectors are required to all be the same length.
+    V.Vector (VS.Vector Double) ->
+    -- | The subject weights. It is required that the length of this vector is equal to the length of each of the vectors in the design matrix.
+    VS.Vector Double ->
+    -- | Indicators for whether each of the covariates should be centered and scaled. It is required that the length of this vector is equal to the number of vectors in the design matrix.
+    V.Vector ScaleCovariateIndicator ->
+    Either T.Text (V.Vector (VS.Vector Double, Double))
 -- The error handling could be improved to (i) list all of the errors and (ii)
 -- to list the indices of the failing vectors
 centerAndScaleCovs xDesignMatrix weights scaleIndicators =
-  let nCovs = V.length xDesignMatrix
-      nScaleIndicators = V.length scaleIndicators
-      sumWeights = VS.sum weights
-  in if nCovs /= nScaleIndicators
-     then Left "The length of first input is not the same as the length of the third input"
-     else do
-       weightedMeans <- calcWeightedMeans sumWeights xDesignMatrix weights scaleIndicators
-       let covTuples = V.zip5 xDesignMatrix
-                              (V.replicate nCovs weights)
-                              (V.convert weightedMeans)
-                              (V.replicate nCovs sumWeights)
-                              scaleIndicators
-           eitherResults = V.map conditionallyCenterAndScaleCov covTuples
-       if V.any isLeft eitherResults
-       then Left "Not all covariates where able to be centered and scaled"
-       else Right $ V.map (fromRight (VS.singleton 0, 0)) eitherResults
+    let nCovs = V.length xDesignMatrix
+        nScaleIndicators = V.length scaleIndicators
+        sumWeights = VS.sum weights
+     in if nCovs /= nScaleIndicators
+            then Left "The length of first input is not the same as the length of the third input"
+            else do
+                weightedMeans <- calcWeightedMeans sumWeights xDesignMatrix weights scaleIndicators
+                let covTuples =
+                        V.zip5
+                            xDesignMatrix
+                            (V.replicate nCovs weights)
+                            (V.convert weightedMeans)
+                            (V.replicate nCovs sumWeights)
+                            scaleIndicators
+
+                traverse conditionallyCenterAndScaleCov covTuples
```

Comment:
Eventually I'd like to improve this function in the future to provide a list of columns that fail to be scaled for the `Left` case. The way the existing code is structured is meant to tee up that extension, although of course it's pretty suboptimal in that it passes through the vector two or three times. Do you see a way of using `traverse` to collect the information for the `Left` case?

Reply:
I added an initial implementation of what I understand you're looking for. Let me know what you think.